File: go/product_families.go
// this file is @generated
package meteroid

import (
	"context"
	"net/http"
)

// ProductFamiliesListProductFamiliesOptions carries the query and header parameters of
// ProductFamilies.ListProductFamilies.
//
// Optional parameters are pointers; leave them nil to omit them.
type ProductFamiliesListProductFamiliesOptions struct {
	// Sort order. Format: `column.direction`. Allowed columns: `name`, `created_at`. Direction: `asc` or `desc`. Default: `created_at.desc`.
	OrderBy *string

	// Page number (0-indexed)
	Page *int32

	// Number of items per page
	PerPage *int32

	Search *string
}

// ProductFamilies groups the product families operations of the Meteroid API.
type ProductFamilies struct {
	client *Client
}

func (a *ProductFamilies) ListProductFamilies(ctx context.Context, options *ProductFamiliesListProductFamiliesOptions) (*ProductFamilyListResponse, error) {
	req := newRequest(http.MethodGet, "/api/v1/product_families")

	if options != nil {
		if options.OrderBy != nil {
			req.SetQueryParam("order_by", *options.OrderBy)
		}
		if options.Page != nil {
			req.SetQueryParam("page", formatInt(int64(*options.Page)))
		}
		if options.PerPage != nil {
			req.SetQueryParam("per_page", formatInt(int64(*options.PerPage)))
		}
		if options.Search != nil {
			req.SetQueryParam("search", *options.Search)
		}
	}

	var out ProductFamilyListResponse
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (a *ProductFamilies) CreateProductFamily(ctx context.Context, productFamilyCreateRequest ProductFamilyCreateRequest) (*ProductFamily, error) {
	req := newRequest(http.MethodPost, "/api/v1/product_families")

	req.SetJSONBody(productFamilyCreateRequest)

	var out ProductFamily
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Retrieve a single product family by ID or alias.
func (a *ProductFamilies) GetProductFamilyByIdOrAlias(ctx context.Context, idOrAlias string) (*ProductFamily, error) {
	req := newRequest(http.MethodGet, "/api/v1/product_families/{id_or_alias}")
	req.SetPathParam("id_or_alias", idOrAlias)

	var out ProductFamily
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: typescript/src/api/productFamilies.ts
// this file is @generated
import { type ProductFamily, ProductFamilySerializer } from "../models/productFamily";
import {
  type ProductFamilyCreateRequest,
  ProductFamilyCreateRequestSerializer,
} from "../models/productFamilyCreateRequest";
import {
  type ProductFamilyListResponse,
  ProductFamilyListResponseSerializer,
} from "../models/productFamilyListResponse";
import { HttpMethod, MeteroidRequest, type MeteroidRequestContext } from "../request";

export interface ProductFamiliesListProductFamiliesOptions {
  /** Sort order. Format: `column.direction`. Allowed columns: `name`, `created_at`. Direction: `asc` or `desc`. Default: `created_at.desc`. */
  orderBy?: string;
  /** Page number (0-indexed) */
  page?: number;
  /** Number of items per page */
  perPage?: number;
  search?: string;
}

export class ProductFamilies {
  public constructor(private readonly requestCtx: MeteroidRequestContext) {}

  /**  */
  public listProductFamilies(
    options?: ProductFamiliesListProductFamiliesOptions
  ): Promise<ProductFamilyListResponse> {
    const request = new MeteroidRequest(HttpMethod.GET, "/api/v1/product_families");

    request.setQueryParam("order_by", options?.orderBy);
    request.setQueryParam("page", options?.page);
    request.setQueryParam("per_page", options?.perPage);
    request.setQueryParam("search", options?.search);
    return request.send(
      this.requestCtx,
      ProductFamilyListResponseSerializer._fromJsonObject
    );
  }

  /**  */
  public createProductFamily(
    productFamilyCreateRequest: ProductFamilyCreateRequest
  ): Promise<ProductFamily> {
    const request = new MeteroidRequest(HttpMethod.POST, "/api/v1/product_families");

    request.setBody(
      ProductFamilyCreateRequestSerializer._toJsonObject(productFamilyCreateRequest)
    );
    return request.send(this.requestCtx, ProductFamilySerializer._fromJsonObject);
  }

  /** Retrieve a single product family by ID or alias. */
  public getProductFamilyByIdOrAlias(idOrAlias: string): Promise<ProductFamily> {
    const request = new MeteroidRequest(
      HttpMethod.GET,
      "/api/v1/product_families/{id_or_alias}"
    );

    request.setPathParam("id_or_alias", idOrAlias);
    return request.send(this.requestCtx, ProductFamilySerializer._fromJsonObject);
  }
}
